### PR TITLE
Introduce restricted channels

### PR DESCRIPTION
**Pull request by AurelienGasser**
**TL;DR:** Channels can be marked as "restricted" (also called "solo"). Restricted channels can only have one member, else the backend will (deliberately) crash.

---

A channel is configured as restricted if either of these conditions are true in the helm values file:

- the `restricted` field of the channel is set to true
- the name of the channel starts with `solo-`

*Note: In the rest of this doc, "misconfigured" refers to a solo channel with more than one member.*

2 protections are implemented to prevent any operation on a misconfigured solo channel.

### Protection 1

Throw an exception when trying to operate on a misconfigured solo channel

`get_hfc(channel_name)` will 

1. check if `channel_name` refers to a solo channel
2. check the channel's discovery results
3. if the channel has more than one member, throw an exception

Since `get_hfc` is the only entry point to the ledger, it is not possible for a consumer to operate on a misconfigured channel _at all_.

### Protection 2

Crash the readiness/liveness probes if a solo channel is misconfigured

Both the readiness and liveness probes will validate all the solo channels: if any solo channel is misconfigured, the HTTP call will return a 500 error, which will instruct Kubernetes to crash the pod. The exception is shown in the logs.

### Test 1: correct configuration

In `backend-org-1.yaml`, add a new channel in `channels`

```yaml
  - solo-org1:
      restricted: true
      chaincode:
        name: mycc
        version: "1.0"
```

In hlf-k8s `examples/2-orgs-policy-any-no-ca/values/org-1-peer-1.yaml`, add a solo channel

```yaml
- channelName: solo-org1
  channelPolicies: |-
    Readers:
        Type: ImplicitMeta
        Rule: "ANY Readers"
    Writers:
        Type: ImplicitMeta
        Rule: "ANY Writers"
    Admins:
        Type: ImplicitMeta
        Rule: "ANY Admins"
  appPolicies: |-
    LifecycleEndorsement:
        Type: ImplicitMeta
        Rule: "ANY Endorsement"
    Endorsement:
        Type: ImplicitMeta
        Rule: "ANY Endorsement"
    Readers:
        Type: ImplicitMeta
        Rule: "ANY Readers"
    Writers:
        Type: ImplicitMeta
        Rule: "ANY Writers"
    Admins:
        Type: ImplicitMeta
        Rule: "ANY Admins"

  chaincodes:
  - address: network-org-1-peer-1-hlf-k8s-chaincode-mycc.org-1
    policy: "OR('MyOrg1MSP.member')"
    name: mycc
    port: 7052
    version: "1.0"
    image:
      repository: substrafoundation/substra-chaincode
      tag: 0.1.0
      pullPolicy: IfNotPresent

  organizations:
    - { org: MyOrg1, mspid: MyOrg1MSP, configUrl: network-org-1-peer-1-hlf-k8s-config-operator.org-1/config/configOrgWithAnchors.json }
```

The backend works correctly.

### Test 2: incorrect configuration

From Test 1, modify the `solo-org1` configuration to add a new `organization` 

```yaml
organizations:
  - { org: MyOrg1, mspid: MyOrg1MSP, configUrl: network-org-1-peer-1-hlf-k8s-config-operator.org-1/config/configOrgWithAnchors.json }
  - { org: MyOrg2, mspid: MyOrg2MSP, configUrl: network-org-2-peer-1-hlf-k8s-config-operator.org-2/config/configOrgWithAnchors.json }
```

Run `skaffold run` and observe the deluge of errors in the backend.

### CODE REVIEW

File: backend/substrapp/ledger/connection.py [hunk 1]
```diff
@@ -119,11 +119,25 @@ def _get_hfc(channel_name):
...
     return loop, client, user
 
 
+def _validate_channels(channel_name, discovery_results):
+
+    channel = [c for c in settings.LEDGER_CHANNELS if c['name'] == channel_name][0]
+
+    # Esnure solo channels have 1 member at most
+    if channel_name.startswith('solo-') or channel['restricted']:
+        channel_members = list(discovery_results['config']['msps'].keys())
```
Comment (Kelvin-M) [inline backend/substrapp/ledger/connection.py:134]: You can use the `results` from `_deserialize_discovery(results)` which already deserialize peer members  in `results['members']` :)

**Review by Kelvin-M — CHANGES_REQUESTED** [review]
Need to be rebased ! 
It may change the way you set the values and load them :)

File: backend/libs/health_check_middleware.py [hunk 2]
```diff
@@ -18,10 +20,21 @@ def liveness(self, request):
...
         """
         Returns that the server is alive.
         """
+        validate_solo_channels()
         return HttpResponse("OK")
+
+
+def validate_solo_channels():
+    for channel_name in settings.LEDGER_CHANNELS.keys():
+        channel = settings.LEDGER_CHANNELS[channel_name]
```
Comment (Kelvin-M) [inline backend/libs/health_check_middleware.py:36]: Python dictionnary has `.items()` method for that :
```python
for channel_name, channel in settings.LEDGER_CHANNELS.items():
```

File: backend/substrapp/ledger/connection.py [hunk 3]
```diff
@@ -123,11 +123,25 @@ def _get_hfc(channel_name):
...
     _update_client_with_discovery(client, results)
 
     return loop, client, user
 
 
+def _validate_channels(channel_name, discovery_results):
+
+    channel = settings.LEDGER_CHANNELS[channel_name]
+
+    # Esnure solo channels have 1 member at most
```
Comment (Kelvin-M) [inline backend/substrapp/ledger/connection.py:136]: Typo `# Ensure`

File: values/backend-org-1.yaml [hunk 4]
```diff
@@ -89,9 +89,11 @@ extraEnv:
 channels:
   - mychannel:
       chaincode:
+        restricted: false
```
Comment (Kelvin-M) [inline values/backend-org-1.yaml:92]: It looks like there is an extra indentation here as it is the channel which is restricted and not the chaincode :)

**Comment by AurelienGasser** [comment]
It's not currently possible to use the same chaincode on different channels, which is a blocker for this PR.

The next step is to make a PR in hlf-k8s.

**Comment by AurelienGasser** [comment]
A PR on hlf-k8s has been started: https://github.com/SubstraFoundation/hlf-k8s/pull/102

**Review by inalgnu — APPROVED** [review]

File: backend/substrapp/ledger/connection.py (hunk 1 above)
Comment (AurelienGasser) [reply backend/substrapp/ledger/connection.py:138]: I can't use that method because the `members` value contains the peers that have **joined** the channel (in this case only peer-1)

```
[[{'mspid': 'MyOrg1MSP', 'id_bytes': '-----BEGIN CERTIFICATE-----\nMIIC0DCCAnegAwIBAgIUA0DXmM3Y0oH+ZgQuQiXrIZlcs04wCgYIKoZIzj0EAwIw\nbjELMAkGA1UEBhMCRlIxGTAXBgNVBAgTEExvaXJlLUF0bGFudGlxdWUxDzANBgNV\nBAcTBk5hb │nRlczEQMA4GA1UEChMHU3Vic3RyYTEPMA0GA1UECxMGRmFicmljMRAw\nDgYDVQQDEwdyY2FPcmcxMB4XDTIwMTAxNDE2MDQwMFoXDTIxMTAxNDE2MDkwMFow\nWjELMAkGA1UEBhMCVVMxFzAVBgNVBAgTDk5vcnRoIENhcm9saW5hMRQwEgYDVQQK\nEwtIeXBlcmxlZGdlcjENMAsGA1UECxMEcGV │lcjENMAsGA1UEAxMEdXNlcjBZMBMG\nByqGSM49AgEGCCqGSM49AwEHA0IABKdE5vumVaXCRq6gTDcBMXnkwpz3cCAhc7fo\nH4fKdk8lUdK/z6Q+Bn9p3f5i8Nzg36IzdMBOfMygmAsiowkjKFqjggEFMIIBATAO\nBgNVHQ8BAf8EBAMCB4AwDAYDVR0TAQH/BAIwADAdBgNVHQ4EFgQUz8uyUuFUi │Pxn\nVXw0hlTCmhRQnvUwHwYDVR0jBBgwFoAUZqHx0bgfq6o42oD4tGd9XOnP71cwSgYD\nVR0RBEMwQYI/bmV0d29yay1vcmctMS1wZWVyLTEtaGxmLWs4cy1lbnJvbGxtZW50\nLW9wZXJhdG9yLTY0ZmY3OTZiN2Z3NnM5MFUGCCoDBAUGBwgBBEl7ImF0dHJzIjp7\nImhmLkFmZmlsaWF0aW9uI │joiIiwiaGYuRW5yb2xsbWVudElEIjoidXNlciIsImhm\nLlR5cGUiOiJwZWVyIn19MAoGCCqGSM49BAMCA0cAMEQCIFxvWtIazap7Cl74yVi2\nPhWju2WqpQ5c20duPC0f+R8gAiAcd1OaUBRNWq2nBbMPQsfIwbCRZzvjm+OvPzBh\nnWyopg==\n-----END CERTIFICATE-----\n', 'ledger │_height': 3, 'chaincodes': [{'name': '_lifecycle', 'version': '1'}], 'endpoint': 'network-org-1-peer-1-hlf-peer.org-1:7051'}], []]
```

While `discovery_results['config']['msps'].keys()` contains all the entities allowed on the channel:

```
['MyOrdererMSP', 'MyOrg2MSP', 'MyOrg1MSP']    
```

We want to raise an error when the channel is configured to allow multiple users (even if only one user has joined the channel)

Comment (Kelvin-M) [reply backend/substrapp/ledger/connection.py:138]: Okay, thanks 👍

**Review by Kelvin-M — APPROVED** [review]

**Comment by AlexandrePicosson** [comment]
I think you forgot to bump the chart.

**Comment by Kelvin-M** [comment]
@AlexandrePicosson Yes !
@AurelienGasser Can you do the bump, like this I can merge it quickly !
